shell/platform/android/io/flutter/plugin/editing/SpellCheckPlugin.java

```diff
@@ -0,0 +1,138 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.plugin.editing;
+
+import android.view.textservice.SentenceSuggestionsInfo;
+import android.view.textservice.SpellCheckerSession;
+import android.view.textservice.SuggestionsInfo;
+import android.view.textservice.TextInfo;
+import android.view.textservice.TextServicesManager;
+import androidx.annotation.NonNull;
+import io.flutter.embedding.engine.systemchannels.SpellCheckChannel;
+import io.flutter.plugin.localization.LocalizationPlugin;
+import java.util.ArrayList;
+import java.util.Locale;
+
+/**
+ * {@link SpellCheckPlugin} is the implementation of all functionality needed for spell check for
+ * text input.
+ *
+ * <p>The plugin handles requests for spell check sent by the {@link
+ * io.flutter.embedding.engine.systemchannels.SpellCheckChannel} via sending requests to the Android
+ * spell checker. It also receives the spell check results from the service and sends them back to
+ * the framework through the {@link io.flutter.embedding.engine.systemchannels.SpellCheckChannel}.
+ */
+public class SpellCheckPlugin
+    implements SpellCheckChannel.SpellCheckMethodHandler,
+        SpellCheckerSession.SpellCheckerSessionListener {
+
+  private final SpellCheckChannel mSpellCheckChannel;
+  private final TextServicesManager mTextServicesManager;
+  private SpellCheckerSession mSpellCheckerSession;
+
+  // The maximum number of suggestions that the Android spell check service is allowed to provide
+  // per word.
+  // Same number that is used by default for Android's TextViews.
+  private static final int MAX_SPELL_CHECK_SUGGESTIONS = 5;
+
+  public SpellCheckPlugin(
+      @NonNull TextServicesManager textServicesManager,
+      @NonNull SpellCheckChannel spellCheckChannel) {
+    mTextServicesManager = textServicesManager;
+    mSpellCheckChannel = spellCheckChannel;
+
+    mSpellCheckChannel.setSpellCheckMethodHandler(this);
+  }
+
+  /**
+   * Unregisters this {@code SpellCheckPlugin} as the {@code
+   * SpellCheckChannel.SpellCheckMethodHandler}, for the {@link
+   * io.flutter.embedding.engine.systemchannels.SpellCheckChannel}, and closes the most recently
+   * opened {@code SpellCheckerSession}.
+   *
+   * <p>Do not invoke any methods on a {@code SpellCheckPlugin} after invoking this method.
+   */
+  public void destroy() {
+    mSpellCheckChannel.setSpellCheckMethodHandler(null);
+
+    if (mSpellCheckerSession != null) {
+      mSpellCheckerSession.close();
+    }
+  }
+
+  @Override
+  public void initiateSpellCheck(@NonNull String locale, @NonNull String text) {
+    performSpellCheck(locale, text);
```

Comment:
Handle case when there's a pending `initiateSpellCheck`.  Maybe, call `result.error` or `result.success` with an empty list. The general idea is that while the framework is waiting on a spellcheck request, a new request is issued, so the older one needs to be dropped.

Reply:
The approach of dropping older requests has a similar fault to the approach I pursued previously of saving the result and text awaiting a response to that result in the `SpellCheckChannel` class, which is that we cannot cancel requests to the Android spell checker once they have been submitted.

We already knew that, but our assumption was that if the `SpellCheckPlugin` has detected a result awaiting a response when another request to initiate spell check has come in, then that request will not reach the `onGetSentenceSuggestions(...)` callback, meaning that the Android spell checker consumed the request. The consequence of assuming such is that sometimes the plugin will attempt to respond to a new request with old spell check results. Depending on how much the text has been modified between the time those old results come in and the time that the new request to spell check was submitted, those old results could lead to some hard-to-handle errors on the framework side.

Given all of that, I decided to have the plugin only handle one request at a time by dropping new requests when an old one is still pending. This approach may lead to results lagging slightly behind, but (i) the spell check results in any given response will correspond to the text in that response, (ii) the spell check results will be returned in chronological order, and (iii) any impacts of these results lagging behind can be handled by the logic on the framework side. This is all because with this approach, we do not have to battle the unpredictable behavior of the Android spell checker. I believe also that in further iterations, other speedup techniques can be more easily explored given these guarantees.

Reply:
Does this approach seem to work with decent performance in practice?  Maybe this is the most solid approach in the real world even though it's not the fastest theoretically.

Reply:
Also, is there ever a risk of the spell checker never responding and locking up or is that another thing that we shouldn't worry about in practice?

Reply:
> Does this approach seem to work with decent performance in practice? Maybe this is the most solid approach in the real world even though it's not the fastest theoretically.

Yes, exactly. From playing around with it (paired with making the corrections on the framework side if the process of returning results lag behind), I don't see any difference. If you are open to playing around with it, too, that'd be great!

> Also, is there ever a risk of the spell checker never responding and locking up or is that another thing that we shouldn't worry about in practice?

I think we can rely on Android for this; they seem to be handling this kind of thing [here](https://android.googlesource.com/platform/frameworks/base/+/master/core/java/android/view/textservice/SpellCheckerSession.java#453).